# Student Performance Indicator

**We are going to follow the following steps:**
1. Understanding Problem Statement
1. Data Collection
1. Data Cleaning
1. EDA & Feature Engineering
1. Data Preprocessing
1. Basic Model Training
1. Basic Model Evaluation

## 1. Understanding Problem Statement

The project aims to understand how the feature and variables like Gender, Ethnicity, Parental level of education, Lunch and Test prepartaion couse etc., can affect a student's performance which is measured by test scores. Hence, Develop a model that can predict the performance of a student by using various features.

## 2. Data Collection

**Dataset Source**  
_https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977_

**Dataset Description**  
Columns: `8`, Rows: `1000`  

Columns  
`Gender` :  Sex of students `=>` [male, female]  
`Race/Ethnicity` :  Ethnicity of students `=>` [groups : a, b, c, d, e]  
`Parental level of education` :  Parent's final education `=>` [bachelor's degree, some college, master's degree, associate's degree, high school]  
`Lunch` :  Having lunch before test `=>` [standard, free/reduced]  
`Test preparation course` :  Status of completion of test `=>` [complete, not complete]  
`Math score` :  Score obtained in Mathematics  
`Reading score` :  Score obtained in Reading  
`Writing score` :  Score obtained in Writing  

## 3. Data Cleaning

#### Importing Libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

#### Importing Data

In [2]:
df = pd.read_csv("../data/student.csv")

#### First View (Top 5 Records)

In [3]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [4]:
df.shape

(1000, 8)

### Data Checks To Peform  

* Check missing values
* Check duplicates
* Check data type
* Check number of unique values i.e., Identify categorical variables
* Check basic statistics of data
* Check various categories present in categorical column (if any)

#### 3.1 Check missing values

In [5]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

#### 3.2 Check duplicates

In [6]:
df.duplicated().sum()

0

#### 3.3 Check data type

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


#### 3.4 Check number of unique of values

In [8]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

#### 3.5 Check basic statistics

In [9]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### 3.6 Check categories in categorical column

In [10]:
print("Categories in different Categorical Features are - \n")

categorical_features = df.columns[df.dtypes == "object"]
numerical_features = df.columns[df.dtypes != "object"]

for category in categorical_features:
    print(f"Categories in '{category}' variable are (Total: {df[category].nunique()})")
    print(df[category].unique(), "\n")

Categories in different Categorical Features are - 

Categories in 'gender' variable are (Total: 2)
['female' 'male'] 

Categories in 'race_ethnicity' variable are (Total: 5)
['group B' 'group C' 'group A' 'group D' 'group E'] 

Categories in 'parental_level_of_education' variable are (Total: 6)
["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school'] 

Categories in 'lunch' variable are (Total: 2)
['standard' 'free/reduced'] 

Categories in 'test_preparation_course' variable are (Total: 2)
['none' 'completed'] 



### Observations & Insights  

* There are NO missing values in dataset.
* There are NO duplicates in dataset.
* Data type of 3/8 Columns is `integer` type and 5/8 columns is `object`.
* Number of unique values  
  * Gender - 2
  * Race/Ethnicity - 5
  * Parental level of education - 6
  * lunch - 2
  * Test preparation course - 2  
  Hence, category count in each categorical column is also low.
* Insights from statistics
  - All means are very close to each other i.e., between 66 and 68.05
  - All standard deviations are also close i.e., between 14.6 and 15.19
  - While there is a minimum score 0 for math, for writing minimum score is 0 which greater than math and for reading score it is much higher with value equal to 17.